internal/server/stat_point.go

```diff
@@ -200,47 +200,113 @@ func (s *Server) GetStatCollection(
+						count := int(c)
+						if count > mostCount.count || (count == mostCount.count && date > mostCount.date) {
+							mostCount = dateCount{date, count}
+						}
+						if date > latestDate.date || (date == latestDate.date && count >= latestDate.count) {
+							latestDate = dateCount{date, count}
+						}
+					}
 				}
 			}
+			// Lastest date is always needed.
+			sv2dates[sv] = []string{latestDate.date}
+			// The date with most place count is needed too.
+			if mostCount.count > latestDate.count {
+				sv2dates[sv] = append(sv2dates[sv], mostCount.date)
+			}
+		}
+	} else {
+		for _, sv := range statVars {
+			sv2dates[sv] = []string{date}
 		}
-		// Construct BigTable row keys.
-		rowList, keyTokens = buildStatCollectionKey(
-			parentPlace, childType, date, statVarObject)
-	}
-	cacheData, err := readStatCollection(ctx, s.store, rowList, keyTokens)
-	if err != nil {
-		return nil, err
 	}
-	for sv, data := range cacheData {
-		if data != nil {
-			cohorts := data.SourceCohorts
-			sort.Sort(SeriesByRank(cohorts))
-			result.Data[sv] = cohorts[0]
+
+	// Get a reverse sorted date list like: [2019, 2017, 2016-01]
+	dateSet := map[string]struct{}{}
+	for _, dates := range sv2dates {
+		for _, date := range dates {
+			dateSet[date] = struct{}{}
 		}
 	}
-	for sv := range result.Data {
-		if result.Data[sv] == nil {
-			result.Data[sv] = &pb.SourceSeries{}
+	dateList := []string{}
+	for date := range dateSet {
+		dateList = append(dateList, date)
+	}
+	sort.Sort(sort.Reverse(sort.StringSlice(dateList)))
+
+	// Query the cache from the most recent date. The order is important as the
+	// source with older date is only used when it has better place coverage.
+	// TODO(shifucun): Parallel the BT query.
+	for _, queryDate := range dateList {
+		rowList, keyTokens := buildStatCollectionKeySvObs(
+			parentPlace, childType, queryDate, statVars)
+		statCache, err := readStatCollection(ctx, s.store, rowList, keyTokens)
+		if err != nil {
+			return nil, err
+		}
+		for sv, data := range statCache {
+			if data == nil {
+				continue
+			}
+			for _, date := range sv2dates[sv] {
+				// Only process when the query date is used for this stat var.
+				if date != queryDate {
+					continue
+				}
+				cohorts := data.SourceCohorts
+				sort.Sort(SeriesByRank(cohorts))
+				// Iterate from the ranked order and find the source with most places.
+				maxSize := 0
+				var usedCohort *pb.SourceSeries
+				for _, cohort := range cohorts {
+					currSize := len(cohort.Val)
+					if currSize > maxSize {
+						maxSize = currSize
+						usedCohort = cohort
+					}
+				}
+				// Add the cohort to result
+				if result.Data[sv].Val == nil {
+					result.Data[sv] = usedCohort
+				} else {
+					// If the result is populated already, then a latest observation
+					// has been added. For this cohort (with an older date), only need
+					// to add the places that is not present yet.
+
+					// TODO(boxu): In this case, the result contains multiple sources,
```

Comment:
this will end up changing the shape of the response in a backwards incompatible way. we should check if there are other callers of this api and version it, or create a new endpoint?

Reply:
This API has not been used yet. I updated the response to include multiple source metadata now.